### PR TITLE
Job Improvement Project - Warlock

### LINKED ISSUE #2128: Skill Tetra Vortex doesn't work when cast by a monster
rAthena Hash: 4a004ad
Client Date: 2017-01-25RagexeRE
Server Mode: Renewall

**Description of Issue**: Skill Tetra Vortex doesn't work in a mob

**Result**: When a monster cast the skill Tetra Vortex in any target, nothing happens. Apparently the skill fails.

**How to Reproduce**: Summon a monster like Fenrir (2564) and wait for it cast the skill on yourself.

### LINKED ISSUE #4706: Job Improvement Project - Warlock
* **rAthena Hash**: N/A

* **Client Date**: N/A

* **Server Mode**: Renewal

* **Description of Issue**: 
  * Result:  Several skill effects were updated for Warlock.
  * Expected Result:  Update these to comply with kRO.
  * How to Reproduce:  Usage of any Warlock skill.
  * Official Information: http://ro.gnjoy.com/news/update/View.asp?seq=239&curpage=1  https://www.divine-pride.net/forum/index.php?/topic/3723-kro-classes-improvement-project/&do=findComment&comment=5902
  

* **Modifications that may affect results**: 
  
   None

This Warlock Update was way back than GX update, it was on 2019.07.17 so, it would be great to have that before future updates.
Thx for the grat work ~!

### PR DESCRIPTION
**Pull request by aleos89**
* **Addressed Issue(s)**: Fixes #4706 and fixes #2128

* **Server Mode**: Renewal

* **Description of Pull Request**: 
  * kRO Changelog: http://ro.gnjoy.com/news/update/View.asp?seq=239&curpage=1

Todo:
- [ ] Add support for int32 item ID. The new Warlock Spellbooks are temporarily disabled for now.

### CODE REVIEW

**Comment by LordWhiplash** [comment]
Hi, thanks for updating this PR!
I've noticed some problems:

- WL_SUMMONFB, WL_SUMMONBL, WL_SUMMONWB and WL_SUMMONSTONE are not working (nothing happens and error in mapserv.bat)
image

- WL_HELLINFERNO is doing damage only on target (no Splash), and the shadow damage should be split in 3 hits. (currently the skill is aplying 2 hits)

**Comment by Badarosk0** [comment]
These two skills now have an icon.

```
 StatusIconChangeTable[SC_MISTY_FROST] = EFST_MISTY_FROST;
 StatusIconChangeTable[SC_MAGIC_POISON] = EFST_MAGIC_POISON;

StateIconList[EFST_IDs.EFST_MISTY_FROST] = {
	haveTimeLimit = 1,
	posTimeLimitStr = 2,
	descript = {
		{ "Freezing", COLOR_TITLE_DEBUFF },
		{ "%s", COLOR_TIME }
	}
}
StateIconList[EFST_IDs.EFST_MAGIC_POISON] = {
	haveTimeLimit = 1,
	posTimeLimitStr = 2,
	descript = {
		{ "Magical Addiction", COLOR_TITLE_DEBUFF },
		{ "%s", COLOR_TIME },
		{ "Decreases resistance against all elemental attacks by 50%." }
	}
}
```

File: src/map/skill.cpp [hunk 1]
```diff
@@ -1701,20 +1701,12 @@ int skill_additional_effect(struct block_list* src, struct block_list *bl, uint1
 	case AB_ADORAMUS:
 		sc_start(src,bl, SC_ADORAMUS, skill_lv * 4 + (sd ? sd->status.job_level : 50) / 2, skill_lv, skill_get_time2(skill_id, skill_lv));
 		break;
-	case WL_CRIMSONROCK:
-		sc_start(src,bl, SC_STUN, 40, skill_lv, skill_get_time(skill_id, skill_lv));
-		break;
 	case WL_COMET:
+		sc_start(src, bl, status_skill2sc(skill_id), 50, skill_lv, 10000); // !TODO: What's the rate?
```
Comment (LordWhiplash) [inline src/map/skill.cpp:1705]: Sigma confirmed to me the rate is 100%
image

**Comment by Balferian** [comment]
Frosty Misty splash area should be 3 cells

**Comment by Badarosk0** [comment]
Reading Spellbook
- The skill should now be Passive.
- Enables player to use new type of spellbooks. The player can click on spellbooks to memorize skill directly.
Currently, if click on spellbooks nothing happens. "I changed to custom item ID minor to test this behavior."

WL_JACKFROST
- Add range 11 when cast the skill.

WL_FROSTMISTY
- Add range 11 when cast the skill.
- The effect of the skill does not appear in the place where the skill was cast. (I don't know if this will be changed on the client, in kRO it is showing the effect of the skill in the cast location and not on the user)

Summon Fire Ball / Summon Water Ball / Summon Ball Lightning / Summon Stone
- If you have 2 active Water Balls and then cast level 2 Summon Fire Ball. 2 Water Balls will be replaced with 5 Fire Balls.
Currently the spheres are not overlapping. The following message is appearing (Cannot summon spheres anymore)

Hell Inferno
- The first damage must be fire and does not have the fire element.

**Comment by tumemmac** [comment]
Summon Fire Ball / Summon Water Ball / Summon Ball Lightning / Summon Stone
- No Summon Ball on every Lv.1

**Comment by teededung** [comment]
Magic Poison has 20 seconds of effect and affects monsters.

**Comment by Balferian** [comment]
@aleos89 I didn’t put it right, the radius should be **SplashArea: 3**
image

**Comment by Dekukun** [comment]
image

The skill still gives problems, any way to fix that?

File: db/re/skill_db.yml [hunk 2]
```diff
@@ -18025,25 +18025,26 @@ Body:
...
+        Count: 3
       - Level: 4
-        Count: -6
+        Count: 4
       - Level: 5
-        Count: -7
+        Count: 5
     Element: Water
-    SplashArea: 9
+    SplashArea: 1
```
Comment (Balferian) [inline db/re/skill_db.yml:18047]: Area should be **SplashArea: 3**

**Comment by aleos89** [comment]
@Badarosk0 Reading Spell Book: Did you make sure to adjust the `ITEMID_WL_MB_SG` to your new item ID?
Hell Inferno: Please see it was fixed prior in 0a063a2.

**Comment by ecdarreola** [comment]
Nothing happens when using the spellbooks (using dummy ItemID)

**Comment by tumemmac** [comment]
Release
Nothing happen on Release Lv.1,  when used summon ball on every lv. or doesn't use.
Thank you guys for hard work. :)

**Comment by aleos89** [comment]
Release level 1 has a new effect. It only casts a memorized spell (from Reading Spellbook).

File: src/map/skill.cpp [hunk 3]
```diff
@@ -15949,21 +15921,27 @@ bool skill_check_condition_castbegin(struct map_session_data* sd, uint16 skill_i
...
-					}
+				}
 
-				if( j < 4 ) { // Need 4 spheres minimum
-					clif_skill_fail(sd,skill_id,USESKILL_FAIL_LEVEL,0);
+				// Cast requirement
+				if (skill_id == WL_TETRAVORTEX)
+					i = 4;
+				else if (skill_id == WL_RELEASE)
+					i = 1;
```
Comment (Balferian) [inline src/map/skill.cpp:15937]: All skills except Tetra vortex don`t need spheres for cast skill.

**Comment by ecdarreola** [comment]
Spellbook item can now be used but release 1 level doesn't release the skill. 

Also, the item is consumed when it should not be consumed as per gif. http://imgc1.gnjoy.com/ufile/common/2019/07/16/122404_MT8EN1Lp.gif

**Comment by aleos89** [comment]
I had tested with Release level 1 and it worked with the book I used.

For the item consumption, did you add the flag 32 to the item_flag db?

**Comment by ecdarreola** [comment]
> I had tested with Release level 1 and it worked with the book I used.
> 
> For the item consumption, did you add the flag 32 to the item_flag db?

Forgot the item_flag, now its not consuming. However the level 1 release doesn't work after using the book. It doesn't release anything. What am I missing?

Did the following:
1. Modified itemID to dummyID
2. itemdb.hpp to reflect itemID (dummy)
3. Recompile

**Comment by aleos89** [comment]
> 
> 
> Reading Spell Book: Did you make sure to adjust the `ITEMID_WL_MB_SG` to your new item ID?
> Hell Inferno: Please see it was fixed prior in [0a063a2](https://github.com/rathena/rathena/commit/0a063a2461a772ed502a6cb8bf341b6d8f1d2ae6).

Did you adjust this constant as well?

**Comment by ecdarreola** [comment]
> > Reading Spell Book: Did you make sure to adjust the `ITEMID_WL_MB_SG` to your new item ID?
> > Hell Inferno: Please see it was fixed prior in [0a063a2](https://github.com/rathena/rathena/commit/0a063a2461a772ed502a6cb8bf341b6d8f1d2ae6).
> 
> Did you adjust this constant as well?

Yes on itemdb.hpp I also change it to e.g. 
`ITEMID_WL_MB_SG						= 60065,`
ItemID and item_flag.

There's an animation of it being used (successful) but release doesn't release the skill.

**Comment by aleos89** [comment]
Did you also adjust db/re/spellbook_db.yml?

**Comment by LordWhiplash** [comment]
I've made some tests on all skills (including Reading Spellbook with dummy ID), and here are my results:

1. Reading Spellbook should have cast time even if it is now used by item (i believe it maintains the same as before changes, since patch notes says nothing about cast changes). (this can be seen here http://imgc1.gnjoy.com/ufile/common/2019/07/16/122404_MT8EN1Lp.gif)
2. Jack Frost when used by Release, continues to have the animation around the user (the damage seems to be applied correctly around the target, i don't know if this will be fixed when on client release, but i thought it is better to get this reported)
3. I could not apply Magic Poison on any target, no damage difference after Comet (casted and released)
4. Tetravortex when used by Release can be used even with no sphere (supposed to be able to use only with 5 spheres?)
5. Tetravortex when used by Release make some weird annimations occour (black rectangles on target)
image
(Tetravortex casted)
image
(Tetravortex Released)

**Comment by LordWhiplash** [comment]
Some more changes needed:

1. Frost Misty should still apply Freezing status, the changes just added "Frost" status (http://kafra.kr/#!/en/KRO/skillinfo/2203)

2. Frost Misty area of effect should be equals to Jack Frost': 
7x7 (level 1 and 2) / 9x9 (level 3 and 4) / 11x11 (level 5) according to Sigma from Divine Pride (https://www.divine-pride.net/forum/index.php?/topic/3723-kro-jobs-improvement-project/&tab=comments#comment-5902)
image

**Comment by tungbach1990** [comment]
When i use new book i got "[Warning]: skill_castend_nodamage_id: Unknown skill used:5078",
How can i fix it. Thanks

**Comment by tungbach1990** [comment]
When i use new book, i got this error, how can i fix this ? Thanks for help !
image
And nothing happen with release skill

**Comment by tungbach1990** [comment]
> When i use new book, i got this error, how can i fix this ? Thanks for help !
> image
> And nothing happen with release skill

I've checked skilldb, i already have it but i got this error :( help me please.
Sorry for my bad english.

**Comment by LordWhiplash** [comment]
> 
> 
> i dont know if this official or not but i think you can change your balls when you use
> Summon fire ball you will get fire balls , i think when you use summon stone the fire balls will change to stone? i don't know about it ,
> about Frosty Misty i think the skill have an effect for the status like the
> 014401_m4B6N243
> 
> i've tried it in 2020-4-01 client and nothing change but the damage increased .
> 
> and the books type Reading Spellbook on the top of the character
> image

Good catch, this blue effect of Frost Misty targets is missing.
About the summons, when you use level 1 summon fire/water/lightning/stone, as long as you don't have 5 spheres, you can use the level 2 skill to replace all of them, but i'm not sure about the official behavior of been able to replace spheres when you already have all 5... makes sense to me been able to replace them (maybe you missclicked?).
The "Reading Spellbook !!" issue is clientside related, you need to add the new "WL_READING_SB_READING" on your skillinfolist.lua/lub.

One issue i'm not certain, is about the releasing order... i'm not sure on if it supposed to be this way, but the releasing order of memorized spellbooks is looking random...
randomrelease
(Reading order in 3 times recorded was Jack Frost>Crimson Rock>Jack Frost>Crimson Rock, but 3 different orders were released).

Thanks a lot for the fixes @aleos89 !

**Comment by Haydrich** [comment]
> > i dont know if this official or not but i think you can change your balls when you use
> > Summon fire ball you will get fire balls , i think when you use summon stone the fire balls will change to stone? i don't know about it ,
> > about Frosty Misty i think the skill have an effect for the status like the
> > 014401_m4B6N243
> > i've tried it in 2020-4-01 client and nothing change but the damage increased .
> > and the books type Reading Spellbook on the top of the character
> > image
> 
> Good catch, this blue effect of Frost Misty targets is missing.
> About the summons, when you use level 1 summon fire/water/lightning/stone, as long as you don't have 5 spheres, you can use the level 2 skill to replace all of them, but i'm not sure about the official behavior of been able to replace spheres when you already have all 5... makes sense to me been able to replace them (maybe you missclicked?).
> The "Reading Spellbook !!" issue is clientside related, you need to add the new "WL_READING_SB_READING" on your skillinfolist.lua/lub.
> 
> One issue i'm not certain, is about the releasing order... i'm not sure on if it supposed to be this way, but the releasing order of memorized spellbooks is looking random...
> randomrelease
> (Reading order in 3 times recorded was Jack Frost>Crimson Rock>Jack Frost>Crimson Rock, but 3 different orders were released).
> 
> Thanks a lot for the fixes @aleos89 !

thanks for your info.
im gonna try to add new  WL_READING_SB_READING
about replacing im just dont have an information for this im just asking .

**Comment by tumemmac** [comment]
- Release
When using GM Command @addperm skill_unconditional without any Summon Ball and then use Release Lv.2 Map Server Crash (Release Lv.1 is fine).

**Comment by Haydrich** [comment]
when you change your Map like go from Prontera to lhz_dun_n , if you use a book while monsters can't touch you , i think while using a books monster able to touch you?
ezgif-5-30381a7dbc86
i dont know if that right or what ?

**Comment by aleos89** [comment]
@Haydrich The reason is because it seems the `itemskill` bonus was set as to not release the invincible timer if the skill is a self-skill. Need to confirm if this is still a thing on kRO or maybe kRO implemented these books in another fashion.

**Comment by Kvoiid** [comment]
Hi, i got this problem while trying to use the spellbook item and i've tried everything to make it work but nothing happens
image
Any help?

File: src/map/battle.cpp [hunk 4]
```diff
@@ -593,6 +593,13 @@ int64 battle_attr_fix(struct block_list *src, struct block_list *target, int64 d
 				}
 				break;
 		}
+
+		if (tsc->data[SC_MAGIC_POISON])
```
Comment (attackjom) [inline src/map/battle.cpp:597]: Magic poison  is effect to magic attack only.

File: src/map/skill.cpp [hunk 5]
```diff
@@ -10056,17 +10024,6 @@ int skill_castend_nodamage_id (struct block_list *src, struct block_list *bl, ui
 			clif_skill_fail(sd,skill_id,USESKILL_FAIL_TOTARGET,0);
 		break;
 
-	case WL_FROSTMISTY:
-		clif_skill_nodamage(src,bl,skill_id,skill_lv,1);
-		map_foreachinallrange(skill_area_sub,bl,skill_get_splash(skill_id,skill_lv),BL_CHAR|BL_SKILL,src,skill_id,skill_lv,tick,flag|BCT_ENEMY,skill_castend_damage_id);
-		break;
-
-	case WL_JACKFROST:
-	case NPC_JACKFROST:
```
Comment (attackjom) [inline src/map/skill.cpp:10065]: NPC_Jackfrost should remain.

File: src/map/battle.cpp (hunk 4 above)
Comment (aleos89) [reply src/map/battle.cpp:597]: The description from Sigma says "Target under this effect will take more damage from all property by 50%.". To me, this sounds like weapon, misc, and magic. Or am I interpreting this wrong?

**Comment by Masao87** [comment]
This seems to be working fine so far now, the only thing missing are the spellbooks / increased item ID range, right?

**Comment by qwerty7vp** [comment]
due to 3776bfb merged
this PR is going to merge too ?

File: src/map/battle.cpp (hunk 4 above)
Comment (attackjom) [reply src/map/battle.cpp:605]: You are correct confirm by this pic.

image

**Pull request merged**